Add IBM Cloud infrastructure support

sayalibhavsar opened the pull request:
# Description

This PR adds automated IBM Cloud VSI provisioning with Terraform, including VPC networking, security, floating IPs, and new Ansible roles with burden script integration.

  Key capabilities added:
  - Automated IBM Cloud infrastructure provisioning via Terraform
  - Support for running multiple instance types in parallel without naming conflicts
  - VPC creation with optional reuse of existing VPCs
  - Automated SSH key detection and validation
  - Resource cleanup utility for systematic deletion of IBM Cloud resources
  - Error handling for failed provisioning and terraform state issues


# Before/After Comparison
IBM Cloud Not Supported,  No Parallel Multi-Instance Support, Manual Resource Management:

# Documentation Check

# Clerical Stuff
 This closes #310 

Relates to JIRA: RPOPC-677

sayalibhavsar committed:
Add IBM Cloud infrastructure support

sayalibhavsar committed:
Change config files to use .tar.gz instead of .zip

sayalibhavsar committed:
fixing inconsistent API key handling between ibm_create & ibm_vpc_create

dvalinrh commented on config/coremark_pro_template.yml:
```diff
@@ -1,5 +1,8 @@
 location: https://github.com/redhat-performance/coremark_pro-wrapper/archive/refs/tags
 exec_dir: "coremark_pro-wrapper-2.1/coremark_pro"
 repo_file: "v2.1.tar.gz"
+rhel_pkgs: bc,numactl,perf,git,unzip,perf
```
All of the config files are wrong, this is adding back in the pkgs stuff which we have removed.

dvalinrh commented on bin/ten_of_us.yml:
```diff
@@ -941,7 +968,7 @@
       vars:
         tf_dir: "tf"
-      when: config_info.system_type == "aws" or config_info.system_type == "gcp" or config_info.system_type == "azure"
+      when: config_info.system_type == "aws" or config_info.system_type == "gcp" or config_info.system_type == "azure" or config_info.system_type == "ibm"
```
Lets change this to be 
when: config_info.system_type in ['aws', 'gcp', 'azure', 'ibm']

dvalinrh reviewed (changes requested):
Good first pass.  Various items (expected, large amount of changes).

sayalibhavsar committed:
Address PR #335 review feedback

- Move API key check from ansible role to burden

- Exit if no SSH key found in IBM Cloud

- Use wait_for_ssh role instead of inline wait

- Extract user readiness check into reusable wait_for_user_ready role

- Revert config file changes (removed re-added pkg lines)

- Fix user_parent_home_dir: move after set_user_name to handle root user

- Change delete_volume_on_instance_delete to true in disks.tf

- Remove wrong CPU type check from ibm_vpc_create

- Remove all pbench references

- Remove hardcoded IBM VPC profiles vars file

- Use in list syntax in ten_of_us.yml

- Simplify README API key documentation

sayalibhavsar committed:
Merge main into pr-335, resolve config file conflicts

dvalinrh commented on bin/burden:
```diff
@@ -1453,6 +1499,21 @@ aws_specific_os_version()
+{
+	# Validate IBM Cloud API key is set - Terraform requires it
+	if [[ -z "$IC_API_KEY" ]] && [[ -z "$IBMCLOUD_API_KEY" ]]; then
+		cleanup_and_exit "Error: IBM Cloud API key not set. Export IC_API_KEY or IBMCLOUD_API_KEY" 1
```
This info should be provided by the scenario file (think CPT pipeline).  You may have it as an exported variable from the shell or via the scenario file.  If from the scenario file, we can always issue an export if that is what the IBM cloud requires, else dump it into the ansible file and use it from there.

dvalinrh reviewed (changes requested):
If the user account never becomes ready we need to exit out. Note, make sure every place you are exiting out, will terminate the system if required.

sayalibhavsar committed:
Add system termination on failure and scenario file API key support

dvalinrh reviewed (approved):
LGTM
